### PR TITLE
Do not simplify type name to a type with the EditorBrowsable attribute on it.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/75162

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 1]
```diff
@@ -265,50 +263,54 @@ private static bool TrySimplify(
...
     {
-        replacementNode = null;
-        issueSpan = default;
+        if (!TrySimplifyWorker(expression, semanticModel, out replacementNode, out issueSpan, cancellationToken))
+            return false;
 
-        switch (expression.Kind())
+        // Ensure that replacement doesn't change semantics.
+        var speculationAnalyzer = new SpeculationAnalyzer(expression, replacementNode, semanticModel, cancellationToken);
+        return !speculationAnalyzer.ReplacementChangesSemantics();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs:271]: removed helper method and pulled this to common location.

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 2]
```diff
@@ -414,51 +416,25 @@ private static bool TrySimplifyMemberAccessOrQualifiedName(
...
-                        if (enclosingSymbol is INamedTypeSymbol namedTypeSymbol)
-                        {
-                            if (namedTypeSymbol.TypeArguments.Length != 0)
-                            {
-                                enclosingTypeParametersInsideOut.AddRange(namedTypeSymbol.TypeArguments);
-                            }
-                        }
-
-                        enclosingSymbol = enclosingSymbol.ContainingSymbol;
-                    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs:447]: this entire while loop did absolutely nothing.

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 3]
```diff
@@ -265,50 +263,54 @@ private static bool TrySimplify(
...
-                    var memberAccess = (MemberAccessExpressionSyntax)expression;
-                    if (IsNonRemovablePartOfDynamicMethodInvocation(semanticModel, memberAccess, cancellationToken))
-                    {
-                        return false;
-                    }
+            replacementNode = null;
+            issueSpan = default;
 
-                    if (TrySimplifyMemberAccessOrQualifiedName(memberAccess.Expression, memberAccess.Name, semanticModel, out var newLeft, out issueSpan))
+            switch (expression)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs:283]: view with whitespace off.

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 4]
```diff
@@ -265,50 +263,54 @@ private static bool TrySimplify(
...
+
+                        if (!TrySimplifyMemberAccessOrQualifiedName(memberAccess.Expression, memberAccess.Name, semanticModel, out var newLeft, out issueSpan))
+                            return false;
+
                         // replacement node might not be in it's simplest form, so add simplify annotation to it.
                         replacementNode = memberAccess.Update(newLeft, memberAccess.OperatorToken, memberAccess.Name)
                             .WithAdditionalAnnotations(Simplifier.Annotation);
-
-                        // Ensure that replacement doesn't change semantics.
-                        return !ReplacementChangesSemantics(memberAccess, replacementNode, semanticModel);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs:288]: the two calls to ReplacementChangesSemantics were pulled up and then inlined.

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 5]
```diff
@@ -414,51 +416,25 @@ private static bool TrySimplifyMemberAccessOrQualifiedName(
...
-
-                    if (containingType != null && !containingType.Equals(leftSymbol))
-                    {
-                        if (leftSymbol is INamedTypeSymbol &&
-                            containingType.TypeArguments.Length != 0)
-                        {
+                        // Don't simplify to a base type if it has the EditorBrowsable attribute on it.  This is
+                        // occasionally done in some APIs to have a 'pseudo internal' base type that has functionality,
+                        // which a user is supposed to only access through some 'pseudo public' derived type instead.
+                        if (!containingType.IsEditorBrowsable(hideAdvancedMembers: true, semanticModel.Compilation, includingSourceSymbols: true))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs:431]: new logic.

**Review by akhera99 — APPROVED** [review]

**Pull request merged**